File: src/main/java/me/taubsie/dungeonhub/application/command/commands/BanAllCommand.java
package me.taubsie.dungeonhub.application.command.commands;

import me.taubsie.dungeonhub.application.command.Command;
import me.taubsie.dungeonhub.application.command.CommandParameters;
import me.taubsie.dungeonhub.application.connection.DiscordConnection;
import me.taubsie.dungeonhub.application.enums.EmbedColor;
import me.taubsie.dungeonhub.application.exceptions.InvalidOptionException;
import me.taubsie.dungeonhub.application.service.ApplicationService;
import me.taubsie.dungeonhub.application.service.ProfileModerationService;
import org.javacord.api.entity.message.embed.EmbedBuilder;
import org.javacord.api.entity.permission.PermissionType;
import org.javacord.api.entity.user.User;
import org.javacord.api.event.interaction.SlashCommandCreateEvent;
import org.javacord.api.interaction.SlashCommandOption;
import org.javacord.api.interaction.SlashCommandOptionBuilder;
import org.javacord.api.interaction.SlashCommandOptionType;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionException;

@CommandParameters(name = "ban-all", description = "Uses the profile moderation to ban all given users.",
        enabledForPermissions = PermissionType.BAN_MEMBERS)
public class BanAllCommand extends Command {
    @Override
    protected void executeCommand(SlashCommandCreateEvent slashCommandCreateEvent) {
        User executor = getUser();
        String users = getStringOption("users");
        String reason = getStringOption("reason");

        String[] userArray = users.split(",");
        List<String> errors = new ArrayList<>();

        if (userArray.length < 1) {
            throw new InvalidOptionException("users", "Please provide a comma-separated list of users to ban.");
        }

        respondLater(new CompletableFuture<EmbedBuilder>().completeAsync(() -> {
            for(String userId : userArray) {
                userId = userId.strip();

                try {
                    User user = DiscordConnection.getInstance()
                            .getBot()
                            .getUserById(userId)
                            .join();

                    ProfileModerationService.getInstance().handleUserBan(getServer(), user, executor, reason);
                }
                catch (CompletionException completionException) {
                    errors.add(userId);
                }
            }

            if (errors.isEmpty()) {
                return ApplicationService.getInstance()
                        .getEmbed()
                        .setColor(EmbedColor.POSITIVE.getColor())
                        .setDescription("Successfully banned " + (userArray.length > 1 ? "all " : "") + userArray.length + " user" + (userArray.length > 1 ? "s." : "."));
            } else {
                return ApplicationService.getInstance()
                        .getEmbed()
                        .setColor(EmbedColor.NEGATIVE.getColor())
                        .setDescription("Couldn't ban the following user(s):\n" + String.join(", ", errors));
            }
        }));
    }

    @Override
    public List<SlashCommandOption> getSlashCommandOptions() {
        SlashCommandOption usersOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.STRING)
                .setName("users")
                .setDescription("Comma-separated list of users to ban.")
                .setRequired(true)
                .build();

        SlashCommandOption reasonOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.STRING)
                .setName("reason")
                .setDescription("Reason for the ban.")
                .setRequired(true)
                .setMinLength(2)
                .build();

        return List.of(usersOption, reasonOption);
    }
}

File: src/main/kotlin/me/taubsie/dungeonhub/application/commands/BanAllCommand.kt
package me.taubsie.dungeonhub.application.commands

import dev.kord.common.entity.Permission
import dev.kord.common.entity.Permissions
import dev.kord.common.entity.Snowflake
import dev.kord.core.supplier.EntitySupplyStrategy
import dev.kordex.core.commands.Arguments
import dev.kordex.core.commands.converters.impl.string
import dev.kordex.core.extensions.Extension
import dev.kordex.core.extensions.publicSlashCommand
import me.taubsie.dungeonhub.application.enums.EmbedColor
import me.taubsie.dungeonhub.application.exceptions.InvalidOptionException
import me.taubsie.dungeonhub.application.loader.LoadExtension
import me.taubsie.dungeonhub.application.service.ProfileModerationService
import me.taubsie.dungeonhub.application.service.addEmbed
import me.taubsie.dungeonhub.application.service.color
import net.dungeonhub.i18n.Translations

/**
 * Command to ban all users in a comma-separated list.
 * @see ProfileModerationService
 */
@LoadExtension
class BanAllCommand : Extension() {
    override val name = "ban-all-command"

    override suspend fun setup() {
        publicSlashCommand(::BanAllArguments) {
            name = Translations.Command.BanAll.name
            description = Translations.Command.BanAll.description
            defaultMemberPermissions = Permissions(Permission.BanMembers)
            allowInDms = false

            action {
                respond {
                    val executor = user.asUser()
                    val users = arguments.users.split(",".toRegex()).dropLastWhile { it.isEmpty() }.toList()

                    val errors: MutableList<String> = ArrayList()

                    if (users.isEmpty()) {
                        throw InvalidOptionException(
                            "users",
                            "Please provide a comma-separated list of users to ban."
                        )
                    }

                    for (userId in users) {
                        val user = try {
                            user.kord.getUser(Snowflake(userId.trim()), EntitySupplyStrategy.cachingRest)
                        } catch (exception: Exception) {
                            null
                        }

                        if (user == null) {
                            errors.add(userId)
                            continue
                        }

                        ProfileModerationService.handleUserBan(guild!!.asGuild(), user, executor, arguments.reason)
                    }

                    if (errors.isEmpty()) {
                        addEmbed {
                            color(EmbedColor.Positive)
                            description =
                                "Successfully banned " + (if (users.size > 1) "all " else "") + users.size + " user" + (if (users.size > 1) "s." else ".")
                        }
                    } else {
                        addEmbed {
                            color(EmbedColor.Negative)
                            description = "Couldn't ban the following user(s):\n${java.lang.String.join(", ", errors)}"
                        }
                    }
                }
            }
        }
    }

    inner class BanAllArguments : Arguments() {
        val users by string {
            name = Translations.Command.BanAll.Arguments.Users.name
            description = Translations.Command.BanAll.Arguments.Users.description
            minLength = 6
        }

        val reason by string {
            name = Translations.Command.BanAll.Arguments.Reason.name
            description = Translations.Command.BanAll.Arguments.Reason.description
            minLength = 2
        }
    }
}